sympy/core/add.py

```diff
@@ -263,6 +263,14 @@ def flatten(cls, seq):
                 # x + O(x**2) -> x + O(x**2)
                 if t is not None:
                     newseq2.append(t)
+
+            # 2 + O(0) -> 2
+            new_order_factors = []
+            for o in order_factors:
+                if not o.expr.is_zero:
```

Comment:
I would have tried to handle this in another way: not adding `O(0)` to `order_factors` at all. So the first thing after this line https://github.com/sympy/sympy/blob/008440578d5242db244849b329aed33a93972e15/sympy/core/add.py#L125
could be

    if o.expr.is_zero:
        continue

Reply:
This is a better way of doing it. Thanks @jksuom